src/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -131,6 +131,11 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
                         // can also occur if the consumer calls it directly and just doesn't pass a constant value.
                         regNumber baseReg = node->ExtractTempReg();
                         regNumber offsReg = node->GetSingleTempReg();
+
+                        // Ensure we aren't overwriting op1Reg
+                        assert(baseReg != op1Reg);
```

Comment:
@CarolEidt, you mentioned
>  It's only needed if the tempReg needs to be different from the target. The internal registers never conflict with incoming sources.

Is the second part (never conflict with incoming sources) also true in the case where a source register is the same as the target register?

Reply:
> in the case where a source register is the same as the target register?

Isn't this the same as being RMW?

Reply:
Basically.

But my question is that:
* since `tmpReg` can equal `targetReg` (when `isInternalRegDelayFree=false`)
* and since `opReg` can equal `targetReg` (when `isDelayFree=false`)
* is it possible that `tmpReg` can equal `opReg` (when `opReg == targetReg`)?
  * or is it a hard limitation that `tmpReg != opReg` (based on @CarolEidt's statement)?

Reply:
No, `tmpReg` can't equal `opReg`. Temporary registers are treated as if they are defined before operands are used so they have to be different.

I think I've seen these details described somewhere in the documentation but I can't find that right now.

Reply:
Ah, it wasn't in the documentation, it's in the comment at the start of lsra.cpp: https://github.com/dotnet/coreclr/blob/f1fee6d7b4d06b17d872fdf587a4f87ca69827c2/src/jit/lsra.cpp#L24-L39

Reply:
@mikedn - thanks for adding all clarification (I'm currently out of the country so my responses are slow and off-timezone)